File: scripts/day7-part1-optimized.py
import time

def solve_part1(grid):
    rows = len(grid)
    cols = len(grid[0])
    
    # Find the starting position 'S'
    start_col = grid[0].index('S')
    
    # Simulate tachyon beams using BFS
    beams = [(0, start_col)]
    visited = {(0, start_col)}
    
    while beams:
        new_beams = []
        
        for row, col in beams:
            new_row = row + 1
            if new_row >= rows:
                continue
            
            cell = grid[new_row][col]
            
            if cell == '.':
                key = (new_row, col)
                if key not in visited:
                    visited.add(key)
                    new_beams.append((new_row, col))
            elif cell == '^':
                visited.add((new_row, col))
                
                if col > 0:
                    left_key = (new_row, col - 1)
                    if left_key not in visited:
                        visited.add(left_key)
                        new_beams.append((new_row, col - 1))
                
                if col < cols - 1:
                    right_key = (new_row, col + 1)
                    if right_key not in visited:
                        visited.add(right_key)
                        new_beams.append((new_row, col + 1))
        
        beams = new_beams
    
    # Count activated splitters
    count = 0
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '^' and (r, c) in visited:
                count += 1
    
    return count

# Read input
with open("../inputs/day-7-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part1(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 7 Part 1 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")

File: scripts/day7-part1-optimized.js
const fs = require("fs");

function solvePart1(grid) {
  const rows = grid.length;
  const cols = grid[0].length;

  // Find the starting position 'S'
  let startCol = grid[0].indexOf("S");

  // Simulate tachyon beams using BFS
  let beams = [[0, startCol]];
  const visited = new Set([`0,${startCol}`]);

  while (beams.length > 0) {
    const newBeams = [];

    for (const [row, col] of beams) {
      const newRow = row + 1;
      if (newRow >= rows) continue;

      const cell = grid[newRow][col];

      if (cell === ".") {
        const key = `${newRow},${col}`;
        if (!visited.has(key)) {
          visited.add(key);
          newBeams.push([newRow, col]);
        }
      } else if (cell === "^") {
        visited.add(`${newRow},${col}`);

        if (col > 0) {
          const leftKey = `${newRow},${col - 1}`;
          if (!visited.has(leftKey)) {
            visited.add(leftKey);
            newBeams.push([newRow, col - 1]);
          }
        }

        if (col < cols - 1) {
          const rightKey = `${newRow},${col + 1}`;
          if (!visited.has(rightKey)) {
            visited.add(rightKey);
            newBeams.push([newRow, col + 1]);
          }
        }
      }
    }

    beams = newBeams;
  }

  // Count activated splitters
  let count = 0;
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] === "^" && visited.has(`${r},${c}`)) {
        count++;
      }
    }
  }

  return count;
}

const input = fs
  .readFileSync("../inputs/day-7-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart1(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 7 Part 1 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
